ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/TransactionsForSenderInfo.java

```diff
@@ -37,34 +34,55 @@ void addTransactionToTrack(
       final long nonce, final PendingTransactions.TransactionInfo transactionInfo) {
     synchronized (transactionsInfos) {
       if (!transactionsInfos.isEmpty()) {
-        final long highestNonce = transactionsInfos.lastKey();
-        if (nonce > (highestNonce + 1)) {
-          LongStream.range(highestNonce + 1, nonce).forEach(gaps::add);
+        final long expectedNext = transactionsInfos.lastKey() + 1;
+        if (nonce > (expectedNext) && nextGap.isEmpty()) {
+          nextGap = OptionalLong.of(expectedNext);
         }
       }
       transactionsInfos.put(nonce, transactionInfo);
+      if (nonce == nextGap.orElse(-1)) {
+        findGap();
+      }
     }
   }
 
-  void updateGaps() {
+  void removeTrackedTransaction(final long nonce) {
+    transactionsInfos.remove(nonce);
     synchronized (transactionsInfos) {
-      final Iterator<Long> nonceIterator = transactionsInfos.keySet().iterator();
-      long previousNonce = -1;
-      while (nonceIterator.hasNext()) {
-        final long currentNonce = nonceIterator.next();
-        LongStream.range(previousNonce + 1, currentNonce).forEach(gaps::add);
-        previousNonce = currentNonce;
+      if (!transactionsInfos.isEmpty() && nonce != transactionsInfos.firstKey()) {
+        findGap();
       }
     }
   }
 
-  NavigableMap<Long, TransactionInfo> getTransactionsInfos() {
-    return transactionsInfos;
+  private void findGap() {
+    // find first gap
+    long expectedValue = transactionsInfos.firstKey();
+    for (final Long nonce : transactionsInfos.keySet()) {
+      if (expectedValue == nonce) {
+        // no gap, keep moving
+        expectedValue++;
+      } else {
+        nextGap = OptionalLong.of(expectedValue);
+        return;
+      }
+    }
+    nextGap = OptionalLong.empty();
```

Comment:
I think I got this to fail:
https://github.com/shemnon/besu/compare/noceGaps...RatanRSur:noceGaps

Reply:
We are only looking for gaps within received transactions in that method. First we are at (1,3,max), then we remove the first TX and are at (3,max).  The gap is between 3 and max, so the answer is 4.  Handling the gap between the pending txes and the world state is not calculated here.

Reply:
Got it